Convert multik-api, multik-jvm, multik-default to use multiplaform. Add native and js targets for multik-jvm

luca992 opened the pull request:
Convert use of service loader to use standard reflection (probably wasn't necessary?)

Reflection for discovering Engines on native doesn't work so add an initEnginesProvider function that only has an implementation for native

updating //apply("$rootDir/gradle/publish.gradle") still needs to be done 

There might be some fat jar issues. That probably should be looked into

luca992 commented:
#14 okio's new [filesystem](https://square.github.io/okio/#file-system-examples) api would make allow this for multiplatform also

f7deleon commented:
hey is there an ETA for this?

luca992 commented:
I pulled the v0.0.2 in. And I added JS as a target.

Someone might want to double check my conversions from Java.lang.System.arraycopy()  to using using kotlin's kotlin.collections.copyInto are correct. (the multiplication of array indexes times 2 is confusing)

luca992 commented:
So I was able to get JS passing all the tests after I reified a bunch of functions to determine NDArray data types on JS. I also simplified/fixed loading engines on js and other targets.

Most tests are passing on all targets. (haven't had time to compare if the tests that are failing are also failing on the v0.0.2 branch.

But, it would be nice if you guys consider reviewing this pull request as it's hard to keep merging work into since the source set folders are reorganized with the multiplatform plugin ( src/kotlin/main -> src/kotlin/commonMain)

luca992 commented:
Failing JvmStatisticsTest:
```
test_of_mean_function_on_a_3d_ndarray[js, node]

AssertionError: Expected <[[3, 5.666666666666667],
[5.666666666666667, 7.666666666666667]]>, actual <[[0, 2.3333333333333335],
[0, 3]]>.
```

Failing JvmLinAlgTest:
```
    JvmLinAlgTest. solve_test[jvm]
    JvmLinAlgTest. solve_test[js, node]
    JvmLinAlgTest. solve_test[iosX64]
    JvmLinAlgTest. solve_test[native]
    JvmLinAlgTest. solve_test[js, browser, ChromeHeadless93.0.4577.82, MacOS10.15.7]
    JvmLinAlgTest. test_Schur_decomposition[native]
    JvmLinAlgTest. test_Schur_decomposition[jvm]
    JvmLinAlgTest. test_Schur_decomposition[iosX64]
    JvmLinAlgTest. test_Schur_decomposition[js, node]
    JvmLinAlgTest. test_eigenvalues[jvm]
    JvmLinAlgTest. test_eigenvalues[iosX64]
    JvmLinAlgTest. test_eigenvalues[js, node]
    JvmLinAlgTest. test_eigenvalues[native]
    JvmLinAlgTest. test_plu[js, node]
    JvmLinAlgTest. test_plu[iosX64]
    JvmLinAlgTest. test_plu[native]
    JvmLinAlgTest. test_plu[jvm]
    JvmLinAlgTest. test_plu[js, browser, ChromeHeadless93.0.4577.82, MacOS10.15.7]
    JvmLinAlgTest. test_qr[js, node]
    JvmLinAlgTest. test_qr[native]
    JvmLinAlgTest. test_qr[jvm]
    JvmLinAlgTest. test_qr[iosX64]
    JvmLinAlgTest. test_upper_hessenberg_form[native]
    JvmLinAlgTest. test_upper_hessenberg_form[jvm]
    JvmLinAlgTest. test_upper_hessenberg_form[js, node]
    JvmLinAlgTest. test_upper_hessenberg_form[iosX64]

```

devcrocod commented:
Hi @luca992, @f7deleon

A small update on what we are doing now and when we will start working on the multiplatform.

We are currently preparing for the [release](https://github.com/Kotlin/multik/milestone/1), which will take place in early October. The changes have accumulated enough to be a minor release of 0.1.0, not 0.0.2. 

As for multiplatform, I plan to take this issue in the next 0.2.0 release. We will analyze in detail the need for a multiplatform and the further development of the library. We will return to you for feedback to find out your cases.

@luca992, you did an amazing job.
I watched your pull request at the very beginning. Thank you so much for contributing to Multik, this is important to me. As I said, I'll come back to this after the current release at the end of October and we'll figure out what needs to be done. 
Keep in mind that now the release will include a dependency on Apache Common CSV. In the future, I'm going to get rid of it in favor of my own multiplatform solution, since I was not satisfied with the performance of the existing libraries.

luca992 commented:
> Hi @luca992, @f7deleon
> 
> A small update on what we are doing now and when we will start working on the multiplatform.
> 
> We are currently preparing for the [release](https://github.com/Kotlin/multik/milestone/1), which will take place in early October. The changes have accumulated enough to be a minor release of 0.1.0, not 0.0.2.
> 
> As for multiplatform, I plan to take this issue in the next 0.2.0 release. We will analyze in detail the need for a multiplatform and the further development of the library. We will return to you for feedback to find out your cases.
> 
> @luca992, you did an amazing job.
> I watched your pull request at the very beginning. Thank you so much for contributing to Multik, this is important to me. As I said, I'll come back to this after the current release at the end of October and we'll figure out what needs to be done.
> Keep in mind that now the release will include a dependency on Apache Common CSV. In the future, I'm going to get rid of it in favor of my own multiplatform solution, since I was not satisfied with the performance of the existing libraries.

Thanks for the update. I'll be trying to keep my fork up-to date with the latest, but any help would be appreciated (as I said merging can be annoying as multiplatform uses different source set folders). If the project (or branch) was at least switched over to the multiplatform plugin it would make things easier.

I'm going to try and keep this fork upto date with the latest as have a need for it on mobile and there's no good multi-dim array library on mobile 🙃(eventually I want to convert the native engine to multiplatform too). Also, it would be super cool to have this running on everything jetpack compose can run on too.

Also, adding a dependency on Apache Common CSV can be handled pretty easily on multiplaftorm with platform specific implementations btw.

luca992 commented:
> Hi @luca992 
> Please note that the test build crashes with an [exception](https://buildserver.labs.intellij.net/buildConfiguration/Kotlin_DataScience_Multik_Tests/140910161?)
> 

I don't have permission to see your build logs.
But, I'll rerun locally to check it out

luca992 commented:
@devcrocod should be fixed now. added arm mac targets too

devcrocod commented:
Yes, that fixed the exception.
The tests failed in multik-jvm, pull up the latest changes.

luca992 commented:
yeah. I see all the tests are passing on your branch. So I probably messed something up when porting System.arraycopy() to using using kotlin's kotlin.collections.copyInto, so I guess that will need to be addressed.

erksch commented:
Hey there, I can't wait to use Multik in our multiplatform project Looking forward to getting this in the official version.

luca992 commented:
I think I can look into trying to get those tests passing soon.

devcrocod commented:
Hi @luca992 
I'm back from dataframe and we have planned a new Multik release.
Two important points for the next release are to add multiplatform and get rid of boxing.

First, please transfer the pull request from the master to the v0.2.0 branch.
I'll take a look at your changes in general.
After that, I think I'll come back to you with questions.

luca992 commented:
Okay, cool @devcrocod I moved it to the v0.2.0 branch. I can try and resolve those failing tests and pull in the latest in the next day or two

luca992 commented:
Okay I pulled the latest, and found and fixed bugs I introduced. I realized I forgot to multiply the size to copy by 2 in ComplexFloatArray.copyOf() and ComplexDoubleArray.copyOf(), that fixed the JvmLinAlgTest tests. There was also a weird JS bug in the mean function causing ++ to increment twice when combined with +=, which caused one test to fail.

So, now all tests should be passing! Let me know if you have any questions!

luca992 commented:
Okay, cool. Not my fault. Updating to kotlin 1.6.10 and adding a large js test time out fixes the issues. Now they're all passing!

devcrocod commented on multik-api/src/commonMain/kotlin/org/jetbrains/kotlinx/multik/ndarray/complex/_ComplexArrays.kt:
```diff
@@ -1353,7 +1353,9 @@ public inline fun ComplexFloatArray.copyOfRange(fromIndex: Int, toIndex: Int): C
     require(newLength >= 0) { "$fromIndex > $toIndex" }
     val ret = ComplexFloatArray(newLength)
-    System.arraycopy(this.getFlatArray(), fromIndex, ret.getFlatArray(), 0, min(this.size - fromIndex, newLength) * 2)
+    this.getFlatArray().copyInto(ret.getFlatArray(), 0,fromIndex, toIndex)
```
It looks like it's over-copying.

devcrocod commented on multik-api/src/commonMain/kotlin/org/jetbrains/kotlinx/multik/ndarray/complex/_ComplexArrays.kt:
```diff
@@ -1353,7 +1353,9 @@ public inline fun ComplexFloatArray.copyOfRange(fromIndex: Int, toIndex: Int): C
-    System.arraycopy(this.getFlatArray(), fromIndex, ret.getFlatArray(), 0, min(this.size - fromIndex, newLength) * 2)
+    this.getFlatArray().copyInto(ret.getFlatArray(), 0,fromIndex, toIndex)
+    val endIndex = min(this.size - fromIndex, newLength) * 2
+    this.getFlatArray().copyInto(ret.getFlatArray(), fromIndex,0, endIndex)
```
Here, apparently, I was mistaken, since I didn't multiply `fromIndex` by 2

luca992 replied:
I'm confused. You think this function is correct, or incorrect? Can you explain how the memory is laidout for complex? I'm not even sure why I'm doing two separate copyIntos @devcrocod

devcrocod commented on multik-api/src/commonMain/kotlin/org/jetbrains/kotlinx/multik/ndarray/complex/_ComplexArrays.kt:
```diff
@@ -1501,7 +1504,8 @@ public operator fun ComplexFloatArray.plus(elements: ComplexFloatArray): Complex
     val result = this.copyOf(thisSize + arraySize)
-    System.arraycopy(elements.getFlatArray(), 0, result.getFlatArray(), thisSize * 2, arraySize * 2)
+    val endIndex = thisSize * 2 + arraySize * 2
+    elements.getFlatArray().copyInto(result.getFlatArray(), thisSize * 2,0, endIndex)
```
Probably you can reuse `copyInto` on complex arrays, this will eliminate the need to multiply indices by two and call `flatArray`

devcrocod commented on multik-api/src/commonMain/kotlin/org/jetbrains/kotlinx/multik/ndarray/data/DataType.kt:
```diff
@@ -58,27 +57,30 @@ public enum class DataType(public val nativeCode: Int, public val itemSize: Int,
-                is ComplexFloat -> ComplexFloatDataType
-                is ComplexDouble -> ComplexDoubleDataType
-                else -> throw IllegalStateException("One of the primitive types was expected")
+        public inline fun <reified T> of(element: T): DataType {
```
This function duplicates `ofKClass`.
Can't we do instance-checks?

luca992 replied:
Sometimes the instance checks don't work on JS https://github.com/Kotlin/multik/pull/20#issuecomment-999227811

devcrocod commented on multik-api/src/commonMain/kotlin/org/jetbrains/kotlinx/multik/ndarray/data/MemoryView.kt:
```diff
@@ -767,7 +779,7 @@ public class MemoryViewComplexFloatArray(override val data: ComplexFloatArray) :
 
     override fun equals(other: Any?): Boolean = when {
```
You changed the above "= when" to "return when". Would like to stick to one style, at least in one file. 
Which one doesn't matter. probably better "return when"

devcrocod commented on multik-api/src/commonMain/kotlin/org/jetbrains/kotlinx/multik/ndarray/operations/IteratingNDArray.kt:
```diff
@@ -293,13 +294,13 @@ public inline fun <T, D : Dimension> MultiArray<T, D>.count(predicate: (T) -> Bo
  * Returns a new array containing only distinct elements from the given array.
  */
-public fun <T, D : Dimension> MultiArray<T, D>.distinct(): NDArray<T, D1> = this.toMutableSet().toCommonNDArray()
+public inline fun <reified T: Any, D : Dimension> MultiArray<T, D>.distinct(): NDArray<T, D1> = this.toMutableSet().toCommonNDArray()
```
Here, too, I would like to get rid of all reified types if possible.

luca992 replied:
https://github.com/Kotlin/multik/pull/20#issuecomment-999227811

devcrocod commented on multik-api/src/commonMain/kotlin/org/jetbrains/kotlinx/multik/ndarray/operations/Transformation.kt:
```diff
@@ -88,9 +89,20 @@ public fun <T, D : Dimension> MultiArray<T, D>.repeat(n: Int): D1Array<T> {
     val data = initMemoryView<T>(size * n, dtype)
     if (consistent) {
         when (this.dtype) {
+            DataType.ByteDataType -> this.data.getByteArray()
```
Probably need to write copyInto for the MemoryView. I will create an issue on this.

devcrocod commented on multik-api/src/commonTest/kotlin/org/jetbrains/kotlinx/multik/iterable/IterableNDArrayTest.kt:
```diff
@@ -15,7 +15,7 @@ import kotlin.test.assertTrue
 
     @Test
-    fun `test of function associate`() {
+    fun `test_of_function_associate`() {
```
can write without snake_case.

luca992 replied:
You can. But JS does not allow spaces in the test function names. So I can either do something like this or add a @JsName(...) annotation for each

devcrocod commented on multik-api/src/jsMain/kotlin/org/jetbrains/kotlinx/multik/api/Engine.kt:
```diff
@@ -0,0 +1,12 @@
+
+public actual fun initEnginesProvider(engines: List<Engine>) {
+    engines.forEach {
+        (enginesProvider as HashMap)[it.type] = it
```
In order not to cast the `enginesProvider`, you can do this:
```kotlin
public actual val _enginesProvider : HashMap<EngineType, Engine> = HashMap()
public actual val enginesProvider : Map<EngineType, Engine>
    get() = _enginesProvider
```

devcrocod commented on multik-api/src/jvmMain/kotlin/org/jetbrains/kotlinx/multik/api/Engine.kt:
```diff
@@ -0,0 +1,22 @@
+        return engines
+    }
+
+public actual fun initEnginesProvider(engines: List<Engine>) {
```
I am confused by the initialization of engines.
I'll think about it

luca992 replied:
https://github.com/Kotlin/multik/pull/20#issuecomment-998354816

devcrocod reviewed (changes requested):
@luca992, I reviewed the `multik-api`. I am delighted with the work you have done.

Basically, I am confused by the initialization of the engines.

Have you tried using multiplatform multik in some projects?

luca992 commented on build.gradle.kts:
```diff
@@ -30,22 +28,8 @@ allprojects {
 
 configure(subprojects.filter { it.name !in unpublished }) {
-    apply("$rootDir/gradle/publish.gradle")
+    //apply("$rootDir/gradle/publish.gradle")
```
This still needs to be updated. I couldn't get it to compile last time I tried. I'll take a look again

luca992 commented on gradle.properties:
```diff
@@ -1,10 +1,18 @@
+org.gradle.caching=true
+kotlin.native.cacheKind.linuxX64=static
+kotlin.native.cacheKind.iosArm64=static
+org.gradle.jvmargs=-Xmx5g -XX:+UseParallelGC
```
I just added that to try and improve build speed. https://developer.android.com/studio/build/optimize-your-build
I can remove that.

luca992 commented on multik-api/build.gradle.kts:
```diff
@@ -3,22 +3,89 @@
+    iosSimulatorArm64()
+    iosX64()
+    js {
+        val timeoutMs = "1000000"
```
there was a 2second timeout most of the linalg tests took longer than that

luca992 commented on multik-api/build.gradle.kts:
```diff
@@ -3,22 +3,89 @@
+        browser{
+            testTask {
+                useMocha {
+                    timeout = timeoutMs
```
Honestly no idea. Never really used kotlin JS for anything, I just added JS support because someone asked. I can look. Also, I'm not sure if there's any point to having both browser and nodejs JS targets. I have to look into what the difference is.

luca992 commented on multik-api/src/commonMain/kotlin/org/jetbrains/kotlinx/multik/api/Engine.kt:
```diff
@@ -6,8 +6,8 @@ package org.jetbrains.kotlinx.multik.api
 import org.jetbrains.kotlinx.multik.api.math.Math
-import java.util.*
-import java.util.concurrent.ConcurrentHashMap
+import org.jetbrains.kotlinx.multik.ndarray.complex.ComplexDoubleArray
```
it's unused. Removed it

luca992 commented on multik-api/src/commonMain/kotlin/org/jetbrains/kotlinx/multik/api/Multik.kt:
```diff
@@ -32,6 +33,7 @@ public typealias mk = Multik
  * @property linalg returns the [LinAlg] implementation of the corresponding engine.
  * @property stat returns the [Statistics] implementation of the corresponding engine.
  */
+@ThreadLocal
```
same as the other ThreadLocal

luca992 commented on multik-api/src/commonMain/kotlin/org/jetbrains/kotlinx/multik/api/Multik.kt:
```diff
@@ -52,7 +54,9 @@ public object Multik {
      */
     public fun addEngine(type: EngineType) {
-        _engines.putIfAbsent(type.name, type)
+        if (!_engines.containsKey(type.name)) {
+            _engines[type.name] = type
+        }
     }
```
Looks like that is a java only convenience function.

devcrocod replied:
Yes, you're right

luca992 commented:
> @luca992, I reviewed the `multik-api`. I am delighted with the work you have done.
> 
> Basically, I am confused by the initialization of the engines.
> 
> Have you tried using multiplatform multik in some projects?

SO, basically it works as follows:
`Engine`'s companion object handles loading engines and holding the reference to the default engine.

Loading the engines happens slightly differently per target:

**JVM**

After some debugging`Engine`'s companion object's `init` function appears to be called automatically when when the program is started (not lazily).  Which is defined as:

```
if (enginesProvider.isNotEmpty()) {
    loadEngine()
}
```
`enginesProvider` is defined per target and for jvm I utilize reflection to find which engine classes are available:
```
public actual val enginesProvider : Map<EngineType, Engine> get() {
        val engineClassNames = listOf(
            "org.jetbrains.kotlinx.multik.jvm.JvmEngine",
            "org.jetbrains.kotlinx.multik.jni.NativeEngine",
            "org.jetbrains.kotlinx.multik.default.DefaultEngine")
        val engines = mutableMapOf<EngineType, Engine>()
        engineClassNames.forEach { e ->
            try {
                val instance = Class.forName(e).kotlin.createInstance() as Engine
                engines[instance.type] = instance
            } catch (t: Throwable){ }
        }
        return engines
    }
```
So essentially, the companion object's init function automatically loads the available engines on JVM and sets the default.

**Native and JS**

Reflection is not fully available on non-JVM targets. As far as I am aware of, there isn't a way to locate and instantiate the available engines on those other targets. 

This `Engine`'s companion object's `init` will have no effect on these platforms[1]. Since, you can't really supply engines before this is called.

So, I created a workaround for non-jvm targets with a global `initEnginesProvider` function which is only defined on non-jvm targets.

```
private val _enginesProvider : HashMap<EngineType, Engine> = HashMap()

public actual val enginesProvider : Map<EngineType, Engine>
    get() = _enginesProvider

public actual fun initEnginesProvider(engines: List<Engine>) {
    engines.forEach {
        _enginesProvider[it.type] = it
    }
    Engine.loadEngine()
}
```

So, for non-jvm targets simply supply the global `enginesProvider` with the engines you want at the start of your program. Like `initEnginesProvider(listOf(JvmEngine()))`. It will handle adding engines to the enginesProvider map without duplicates and call `Engine.loadEngine()` to set the default engine.


[1] Disclaimer: For native builds the companion object behavior appears to be different than jvm and js. After debugging`Engine`'s companion object's `init` function on native it appears to be called lazily only when the companion object is called. So, it would actually be possible to set enginesProvider before `Engine`'s companion object's `init` function is called on native. (However, I made the native and js implementations the same for consistency)

luca992 commented on multik-api/src/commonMain/kotlin/org/jetbrains/kotlinx/multik/api/Engine.kt:
```diff
@@ -78,34 +71,35 @@ public abstract class Engine {
-}
+public expect val enginesProvider : Map<EngineType, Engine>
+
+public expect fun initEnginesProvider(engines: List<Engine>)
```
https://github.com/Kotlin/multik/pull/20#issuecomment-998354816

luca992 commented:
@devcrocod 
So for all the reified's I added for JS. It is due to this issue
```
        //example
        val a = mk.ndarrayOf(1.0, 2.2, 3.0)
        val multiArrayIter = (a as MultiArray<*, *>).iterator()
        println(multiArrayIter.next()!!::class)
        println(multiArrayIter.next()!!::class)
        println(multiArrayIter.next()!!::class)
        val ndArrayIter = a.iterator()
        println(ndArrayIter.next()::class)
        println(ndArrayIter.next()::class)
        println(ndArrayIter.next()::class)
```

for js prints:
```
class Int
class Double
class Int
class Double
class Double
class Double
```

Whereas for jvm and native it correctly prints as:
```
class kotlin.Double
class kotlin.Double
class kotlin.Double
class kotlin.Double
class kotlin.Double
class kotlin.Double
```

I think this is due to ImmutableMemoryView storing data as type `any` so that's why I removed the instance-checks. I'm not sure how else to handle it. Any ideas?

luca992 commented:
@devcrocod so this is the js issue I worked around with the reified functions:

```
    inline fun <reified T: Any> printClass(num: T){
        println("printClass num:class ${num::class}")
        println("printClass reified ${T::class}")
    }

    @Test
    fun testtt(){
        val a = listOf(1.0f,2.1f)
        println("a[0] ${a[0]::class}")
        printClass(a[0])

        println("a[1] ${a[0]::class}")
        printClass(a[1])
    }
```

In js outputs:
```
a[0] class Float
printClass num:class class Int
printClass reified class Float
a[1] class Float
printClass num:class class Double
printClass reified class Float
```

where as jvm and native output as expected:
```
a[0] class kotlin.Float
printClass num:class class kotlin.Float
printClass reified class kotlin.Float
a[1] class kotlin.Float
printClass num:class class kotlin.Float
printClass reified class kotlin.Float
```

luca992 commented:
So I managed to revert almost all the `reified`. 

The main problem was that calling in JS builds`Iterable.toCommonNDArray()` causes it to lose type info of the iterable you call it on. So, for example, if `1.0f` was the first element of the iterable when `ndarrayCommon()` tries to get datatype with `DataType.of(elements.first())` it will actually return an `int` data type instead of `float`.

So I made a workaround by creating two implementations of `toCommonNDArray` one which uses `reified` and one which requires a datatype to pass to 
`ndarrayCommon()` in a new optional `dtype` param to force datatype

luca992 commented on multik-api/src/commonMain/kotlin/org/jetbrains/kotlinx/multik/api/CreateNDArray.kt:
```diff
@@ -396,7 +400,7 @@ internal fun <T, D : Dimension> ndarrayCommon(elements: Collection<T>, shape: In
  * @sample samples.NDArrayTest.ndarrayCollections1D
  */
-public fun <T : Number> Multik.ndarray(elements: Collection<T>): D1Array<T> =
+public inline fun <reified T : Number> Multik.ndarray(elements: Collection<T>): D1Array<T> =
```
https://github.com/Kotlin/multik/pull/20#issuecomment-999227811

devcrocod commented:
Hi @occultus73
I remember you asked to add kotlin/js.
Please tell us more about your use case.
What backend are you using?

occultus73 commented:
My boss/client said he wanted the capability to deploy the baysian context bandit library (basically a generic machine learning model) I built on multiplatform with a colleague to JavaScript environments. What exactly he had in mind I'm not sure, but he did express interest in the file io capabilities that come with node.js; as this allows the bandit to save and resume its learning progress in matrix files (I built custom inputs because we didn't have .npy API then).

devcrocod commented:
@luca992 
Thanks for reading the js issue in more detail. Unfortunately, it seems that from the side of the kotlin/js team this will not be solved now, but let's see what they answer.

Using `dtype` is probably the only way for now. It won't help when our data comes from outside, but I think this is a good trade-off. I'll take a look at the related changes.

As for Engines and EnginesProvider. 
I would like to make an equivalent replacement for SPI. Because in the future, multik-native and multik-default will also be multiplatform. I was prompted to see how it was done in ktor:
https://github.com/ktorio/ktor/blob/1.5.4/ktor-client/ktor-client-features/ktor-client-json/common/src/io/ktor/client/features/json/JsonFeature.kt#L24
https://github.com/ktorio/ktor/blob/1.5.4/ktor-client/ktor-client-features/ktor-client-json/ktor-client-serialization/js/src/SerializerInitializer.kt
https://github.com/ktorio/ktor/blob/1.5.4/ktor-client/ktor-client-features/ktor-client-json/ktor-client-serialization/posix/src/SerializerInitializer.kt

It seems to me that for the moment you can leave what you have done.

luca992 commented:
> 



> @luca992 Thanks for reading the js issue in more detail. Unfortunately, it seems that from the side of the kotlin/js team this will not be solved now, but let's see what they answer.
> 
> Using `dtype` is probably the only way for now. It won't help when our data comes from outside, but I think this is a good trade-off. I'll take a look at the related changes.
> 
> As for Engines and EnginesProvider. I would like to make an equivalent replacement for SPI. Because in the future, multik-native and multik-default will also be multiplatform. I was prompted to see how it was done in ktor: https://github.com/ktorio/ktor/blob/1.5.4/ktor-client/ktor-client-features/ktor-client-json/common/src/io/ktor/client/features/json/JsonFeature.kt#L24 https://github.com/ktorio/ktor/blob/1.5.4/ktor-client/ktor-client-features/ktor-client-json/ktor-client-serialization/js/src/SerializerInitializer.kt https://github.com/ktorio/ktor/blob/1.5.4/ktor-client/ktor-client-features/ktor-client-json/ktor-client-serialization/posix/src/SerializerInitializer.kt
> 
> It seems to me that for the moment you can leave what you have done.

Interesting. I thought I tried a similar approach but noticed the init function of the objects weren't always called on all platforms. Looks like they get around that by assigning the object to some global variable. I could try and test that approach again

devcrocod commented:
> I could try and test that approach again

At your discretion.
As I said so far, the current solution is a good place to start.

devcrocod commented on multik-default/build.gradle.kts:
```diff
@@ -1,14 +1,47 @@
-    implementation(project(":multik-jvm"))
-    implementation(project(":multik-native"))
+plugins {
+    kotlin("multiplatform")
```
multik-native is for jvm only, why is multik-default multiplatform then?

luca992 replied:
I would have converted `multik-native` too. However it has a bunch of custom tasks and uses the `cpp-library` plugin. So if you wanted me to do it I would have to do some research on the correct way to handle that. I did convert `multik-default` to using the multiplatform plugin, however it only targets jvm, so basically nothing is different for  `multik-default` I just prepped the gradle config to be able to add more targets

luca992 replied:
I just realized I forgot to move multik-default src/main to src/jvmMain.. now that is fixed and it is correctly built

devcrocod commented on multik-default/src/main/kotlin/org/jetbrains/kotlinx/multik/default/DefaultMath.kt:
```diff
@@ -0,0 +1,112 @@
+ * Copyright 2020-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package org.jetbrains.kotlinx.multik.default
```
This file has changed its location, now it is in `org.jetbrains.kotlinx.multik.default.math`

devcrocod reviewed (changes requested):
@luca992
I looked at other modules and new changes.
Globally, I have no new questions.

I think to do this: you will make all the small edits that remain, after that, I merge your changes. Everything related to publishing, dokka, initializing engines, and js, I will issue it as an issue, if you want, you can take them to work. This will make it easier to pour in the changes and it will be more convenient for me to participate. What do you think?

luca992 commented:
> @luca992 I looked at other modules and new changes. Globally, I have no new questions.
> 
> I think to do this: you will make all the small edits that remain, after that, I merge your changes. Everything related to publishing, dokka, initializing engines, and js, I will issue it as an issue, if you want, you can take them to work. This will make it easier to pour in the changes and it will be more convenient for me to participate. What do you think?

Sounds good to me. 👍 I'm still confused about layout of memory in complex arrays. But if you think what I did was wrong, addressing it as an issue works for me

luca992 commented:
Also, I think I am going to remove the ThreadLocal annotations and just enable the new native memory model. It seems to me that fully adopting the new memory model is what most devs are doing these days. For reference:

https://twitter.com/kpgalligan/status/1474101883686068227?s=20

Dealing with the difficult to resolve concurrent memory issues in the old memory model is annoying to say the least

devcrocod commented on multik-api/src/commonMain/kotlin/org/jetbrains/kotlinx/multik/api/Engine.kt:
```diff
@@ -63,10 +43,23 @@ public abstract class Engine {
      */
     public abstract fun getStatistics(): Statistics
 
+    @ThreadLocal
```
Better to use the new model, as it will become the default soon

devcrocod reviewed (approved):
@luca992 
I'm merging your changes.

As I said, I created issues for a number of problems, including complex arrays. With complex arrays, nothing complicated, I can fix it myself. The principle of complex arrays is simple, the real and imaginary parts are sequentially stored there, that is, for each step it is x2.

There was a problem while running the tests. When calling the `clean` task, I got an exception: `java.lang.NullPointerException: destinationDir must not be null`. I found the corresponding issue in ktor, it will be necessary to upgrade to 1.6.20 where it was corrected.
Otherwise, the library passed the tests, everything is fine.

luca992 commented:
Okay great! Now that it's merged I could take a look at porting multik-native when I get a chance if you guys don't beat me to it.

devcrocod commented:
To port multik-native, you first need to figure out how to build OpenBLAS and bundle it all.
Plus, besides, it will be necessary to check all this on the appropriate platforms.

luca992 commented:
> To port multik-native, you first need to figure out how to build OpenBLAS and bundle it all.
> 
> Plus, besides, it will be necessary to check all this on the appropriate platforms.


Yeah, it looks simple enough to build a shared lib for each target. I'd probably just create a new "libs" folder in the root that contains the shared libs and the headers. Then link it up with kotlin cinterop
